Adjust build configuration to use vendored FluxC and WordPressLoginFlow

wzieba opened the pull request:
Depends on
- #21304 
- #21266 


### Description

This PR replaces the FluxC from an external repository with vendored modules introduced in #21260.

The changes here relate to different areas: static code analysis, build itself or code removal. Please see commits messages for a bit of context.

### Test

Please smoke test the project: focus on upgrade scenarios (installing this version on top of the one that uses FluxC from a repository), think about database migrations and other areas that could be impacted by this change.

```[tasklist]
### Tasks
- [x] verify `test` Gradle tasks (if they run FluxC tests as well)
- [x] perform experiments on how this affects build times ( pbArwn-6Nk-p2 )
- [x] Namespace 'org.wordpress.android.fluxc' is used in multiple modules and/or libraries: :fluxc, org.wordpress:fluxc:trunk-ed60798b4d96ec19863c74b0f525e2e20f4525db. Please ensure that all modules and libraries have a unique namespace. For more information, See https://developer.android.com/studio/build/configure-app-module#set-namespace
- [x] R8: Type org.wordpress.android.fluxc.action.EditorThemeAction is defined multiple times:
- [x] Update test suites tokens
```  

### Before

```mermaid
%%{
  init: {
    'theme': 'neutral'
  }
}%%

graph LR
  subgraph :libs
    :libs:mocks["mocks"]
    :libs:annotations["annotations"]
    :libs:networking["networking"]
    :libs:analytics["analytics"]
    :libs:image-editor["image-editor"]
    :libs:editor["editor"]
    :libs:processors["processors"]
  end
  :WordPress --> :libs:mocks
  :WordPress --> :libs:annotations
  :WordPress --> :libs:networking
  :WordPress --> :libs:analytics
  :WordPress --> :libs:image-editor
  :WordPress --> :libs:editor
  :WordPress --> :libs:processors
  :libs:processors --> :libs:annotations
```


### After

```mermaid
%%{
  init: {
    'theme': 'neutral'
  }
}%%

flowchart LR
  subgraph :libs
    :libs:fluxc-processor["fluxc-processor"]
    :libs:fluxc-annotations["fluxc-annotations"]
    :libs:fluxc["fluxc"]
    :libs:mocks["mocks"]
    :libs:annotations["annotations"]
    :libs:networking["networking"]
    :libs:analytics["analytics"]
    :libs:image-editor["image-editor"]
    :libs:editor["editor"]
    :libs:login["login"]
    :libs:processors["processors"]
  end
  :libs:fluxc-processor --> :libs:fluxc-annotations
  :libs:fluxc --> :libs:fluxc-annotations
  :libs:fluxc --> :libs:fluxc-processor
  :WordPress --> :libs:mocks
  :WordPress --> :libs:annotations
  :WordPress --> :libs:networking
  :WordPress --> :libs:analytics
  :WordPress --> :libs:image-editor
  :WordPress --> :libs:editor
  :WordPress --> :libs:fluxc
  :WordPress --> :libs:login
  :WordPress --> :libs:processors
  :libs:processors --> :libs:annotations
  :libs:login --> :libs:fluxc
style :libs:fluxc fill:#f40,stroke:#333,stroke-width:4px
style :libs:fluxc-processor fill:#f40,stroke:#333,stroke-width:4px
style :libs:fluxc-annotations fill:#f40,stroke:#333,stroke-width:4px
style :libs:login fill:#f40,stroke:#333,stroke-width:4px

```

### Codecov

Includes FluxC and Login as well

<img width="1578" src="https://github.com/user-attachments/assets/9d263347-1dde-4ceb-b14b-d526473d48fe">

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

wzieba commented on .buildkite/commands/run-unit-tests.sh:
```diff
@@ -17,6 +17,16 @@ elif [ "$1" == "image-editor" ]; then
+    test_results_dir="fluxc/build/test-results"
+    test_log_dir="${test_results_dir}/testReleaseUnitTest/*.xml"
+    code_coverage_report="fluxc/build/reports/kover/reportRelease.xml"
+elif [ "$1" == "WordPressLoginFlow" ]; then
+    test_suite=":WordPressLoginFlow:testReleaseUnitTest :WordPressLoginFlow:koverXmlReportRelease"
+    test_results_dir="WordPressLoginFlow/build/test-results"
+    test_log_dir="${test_results_dir}/testReleaseUnitTest/*.xml"
+    code_coverage_report="WordPressLoginFlow/build/reports/kover/reportRelease.xml"
```
There's only one unit test here 😅 I'm not sure if it's worth to run this.

ParaskP7 replied:
Yea, I am not sure as well, but maybe yes, especially if it is cheap, just to keep the configuration consistent on each module, and then to not have to worry about missing that going forward, say we start adding more unit tests for `Login`... 🤷 

PS: Totally unrelated @wzieba , organizational wise, now that we have this chance and just to keep things consistent and avoid confusion/questions down the line:
- Why not moving all these vendored modules into their own folder, even maybe calling it `vendored`. Just like it happens with the `libs` related modules. Actually, I would be totally okay if we move them to this existing `libs` folder, since some of those libs within that folder went through this exact process, them being vendored and all. 🤔 
- Why not renaming the `WordPressLoginFlow` module to plain `login`. 🤔

wzieba replied:
> Why not moving all these vendored modules into their own folder, even maybe calling it vendored.

I tried to reduce git history manipulation to minimum. I also don't see benefits of introducing these modules into a directory: for me, it makes task names longer, instead of `fluxc:assembleRelease` we'd have to run `:libs:fluxc:assembleRelease`. What benefits do you see that moving these 2 modules to a directory would bring?

> Why not renaming the WordPressLoginFlow module to plain login. 🤔

We can rename the project name to `login` without renaming the directory. WDYT?

ParaskP7 replied:
> I tried to reduce git history manipulation to minimum.

👍 

> I also don't see benefits of introducing these modules into a directory: for me, it makes task names longer, instead of fluxc:assembleRelease we'd have to run :libs:fluxc:assembleRelease.

Yea, true, but then again, we rarely run there `assemble` commands on the lib module itself and usually run the main app module assemble command for all, so I don't think this is a problem, having the names longer. 🤷 

> What benefits do you see that moving these 2 modules to a directory would bring?

Not only these 2 modules, but actually all 4 of them, if I recall correctly:
- WordPressLoginFlow
- fluxc-annotations
- fluxc-processor
- fluxc

The benefits I see is mostly organizational wise (as I said earlier), just so that we don't pollute the root level directories with extra (temporar) modules/folders... 🤷 

I also think that naming that folder as `vendored` would also help others understand where and why those modules got moved into the repo. Else, using the existing `libs` folder is perfectly fine as well, and one less folder to worry about.

> We can rename the project name to login without renaming the directory. WDYT?

Hmmm, I don't know, wouldn't that further complicate things, having 2 names for the same things, a project and a folder name. 🤔 I think it is better to stick with `WordPressLoginFlow`, that is, if renaming that to `login` is too troublesome.

wzieba replied:
I see @ParaskP7 👍 What's your preference then: moving these modules to `libs` or creating a new directory: `vendored`? It's worth to note that we already have `vendor` directory created when ruby dependencies are installed

ParaskP7 replied:
`libs` it is then @wzieba , that would have been my preferred approach no matter if the `vendor` directory existed or not, especially since some of the other existing `libs` related modules were vendored in there too. I only suggested `vendored` in case we wanted to separate these two directories, just because `fluxc` is to be replaced at some point with `rs`. Thanks for asking! 🙏

wzieba replied:
Thanks 👍 . Vendored projects are now in `libs` with `WordPressLoginFlow` renamed to simply `login` 🙌

wzieba commented on .buildkite/pipeline.yml:
```diff
@@ -107,6 +107,26 @@ steps:
+          - $CI_TOOLKIT
+          - $TEST_COLLECTOR :
+              <<: *test_collector_common_params
+              api-token-env-name: "BUILDKITE_ANALYTICS_TOKEN_UNIT_TESTS_WORDPRESS"
```
I think the analytics token is fine - I don't think we need to have special token for each test job.

@AliSoftware do you know if the above make sense? Is there a reason to have a unique `api-token-env-name` for each test job in pipeline?

AliSoftware replied:
@mokagio might know more as I think he has more experience in setting up Buildkite Test Analytics, but iinm the way BTA works it uses the token to know which Test Analytics Suite to associate that run with (eg https://buildkite.com/organizations/automattic/analytics/suites/fluxc-android-unit-tests/suite_token)

So the use of different tokens here is not a matter of security or similar but more because we want those different types of tests to be reported in different "Test Suites" in Buildkite, so that the analytics for each don't all end up in the same stats bucket but instead have their own independent graphs and stats in Buildkite.

mokagio replied:
> So the use of different tokens here is not a matter of security or similar but more because we want those different types of tests to be reported in different "Test Suites" in Buildkite, so that the analytics for each don't all end up in the same stats bucket but instead have their own independent graphs and stats in Buildkite.

Yeah, that was the reason. A bit more context in paaHJt-3wn-p2#comment-6208

wzieba replied:
I see, thanks! I've added new test suites and added tokens to CI's `env`. Soon, I update this pipeline.

wzieba commented on WordPressLoginFlow/build.gradle:
```diff
@@ -42,6 +32,36 @@ android {
+    }
+}
+
+ext {
```
Sorry @ParaskP7 ! I hope it's okay for you - I'd rather merge it this way (copy-paste from WordPressLoginFlow) and migrate to versions catalog in a separate PR, to reduce changes in this PR.

ParaskP7 replied:
Of course, this makes total sense to me and thanks for the heads-up @wzieba ! 👍

wzieba commented on build.gradle:
```diff
@@ -89,7 +97,7 @@ allprojects {
         kotlinOptions {
             jvmTarget = JavaVersion.VERSION_1_8
-            allWarningsAsErrors = true
+            allWarningsAsErrors = false
```
Context (internal): pdnsEh-1S9-p2#comment-4233

ParaskP7 replied:
👋 @wzieba !

I recommend reverting this change and instead add the following only on the `fluxc` module to not treat warning as error only at this level and let everything else stay as is. Wdyt?

```
tasks.withType(KotlinCompile).configureEach {
    compilerOptions {
        allWarningsAsErrors = false
    }
}
```

FYI: This is not applicable to `fluxc-annotations` and `fluxc-processor` modules, them being `java` and not `kotlin` module, thus no change needed there. Also, the `login` module was already adhering to this standard, so no change there needed as well.

PS: I am suggesting that because I don't think we'll come back and try to resolve every warning on the `fluxc` level, while in the meanwhile unnecessarily disabling this for all other modules. To my understanding, we should make minimal changes to this `fluxc` module from here onwards, with the end goal of being, to replace this `fluxc` module with `wordpress-rs` as some point.

wzieba replied:
I agree with everything you said here, good points 👍 Adjusted in 952ebd7f2703e49f2e6a9daedeeece0d0a3a1051

wzieba committed:
Remove publish code from vendored modules

wzieba committed:
Build app with vendored FluxC

Update Kotlin to 1.9.25 to align app to use the same version.

wzieba committed:
FluxC: Adjust code to WP checkstyle rules

wzieba committed:
FluxC: Adjust lint configuration, temporarily ignore some issues

wzieba committed:
FluxC: Adjust detekt configuration

Database models have to be mutable due to WellSql constraints

wzieba committed:
FluxC: Adjust unit tests

Add necessary dependencies. DateUtils is a helper class used only in tests. I've copied it manually as we only need a small portion of it, comparing to FluxC original.

ArrayUtilsTest was a test that was testing FluxC's example app code.

wzieba committed:
Login: build app with vendored login library

wzieba committed:
Login: Adjust code to WP checkstyle rules

wzieba committed:
Login: Adjust lint configuration, address easy-to-fix issues

wzieba committed:
Login: Suppress detekt warning

wzieba committed:
Run unit tests for FluxC and Login on CI

wzieba committed:
Remove unused imports

wzieba committed:
Move login library from `REMOTE_LIBRARIES_STRINGS_PATHS` to `LOCAL_LIBRARIES_STRINGS_PATHS` in `localization.rb` script

wzieba committed:
Update documentation, remove unused code related to including FluxC or Login Flow builds

wzieba commented on fastlane/lanes/localization.rb:
```diff
@@ -253,14 +254,6 @@
-      strings_file_path: 'WordPressLoginFlow/src/main/res/values/strings.xml',
-      exclusions: ['default_web_client_id'],
-      source_id: 'login'
-    },
```
👋 @AliSoftware @mokagio WordPress-Login-Flow will be soon part of this repository. Will this change simply work® or do we have to adjust this somewhere else as well?

AliSoftware replied:
Yes. Given you made this change in conjunction with the other change you made on like 247 so that the `strings.xml` from that newly-local Login library are now pulled from the local module instead of the remote repo, I think this should just work® 🙂

wzieba replied:
Thanks!

wzieba commented:
Because there were multiple and unresolvable git conflicts (rebase + changed paths) and commit history was already quite messy, I've decided to start over and force push to this branch. Majority of changes is the same. If you want to see old diff, it's available here: https://github.com/wordpress-mobile/WordPress-Android/compare/47ca6aa47946335b87d62ad4422640ff6b4973db...777dc93c97b67772c46f2d13921e07ee557c3437?diff=unified&w=

wzieba committed:
Remove trailing coma from localization.rb

ParaskP7 commented on libs/fluxc-annotations/build.gradle:
```diff
@@ -1,26 +1,8 @@
 plugins {
     id "java"
```
Suggestion (💡): Let move this to version catalogs as well: `alias(libs.plugins.java)`

```toml
[plugins]
...
java = { id = "java" }
...
```

PS: Same applies for the `fluxc-processor` module.

wzieba replied:
Thanks, but doesn't it feel redundant in this case? We write more code in 2 files instead of simply having `id "java"` in one 🤔 What are the benefits to use catalog and alias in this case?

ParaskP7 replied:
Yea, I understand what you are saying, but for me it is about:
1. Consistency -> Having `alias(libs.plugins.xzy)` everywhere and having everyone adhering to this standard, that is, instead of a random `id "xyz"`. Think constants definition but for builds.
2. Updatability -> Keep that `id` string in one place and one place only, version catalogs for that matter, then, when and if needed, update it in that one place, that is, instead of multiple places.

wzieba replied:
I see, thanks for sharing, but frankly speaking this doesn't convince me:

> Consistency -> Having alias(libs.plugins.xzy) everywhere and having everyone adhering to this standard, that is, instead of a random id "xyz". Think constants definition but for builds.

I don't think `id "java"` is random - it's a standard way of declaring plugins in Gradle. It's already a constant in this sense.

> Updatability -> Keep that id string in one place and one place only, version catalogs for that matter, then, when and if needed, update it in that one place, that is, instead of multiple places.

I can't imagine that `java` plugin will change its name. If this, highly unlikely IMO, scenario happen, we'll have to only update 2 files, so I think we shouldn't worry about it.

Taking both into account, for now I'd not introduce java to version catalog, as I simply don't see benefits and I see downsides: boilerplate code.

ParaskP7 replied:
Totally, up to you, but honestly, for example, if I where to start a new project, I would put `java` into version catalogs and not have that there be different from the rest.

1. About the random thing, I was trying to make an analogy with constants. We've already used to move any magic number or random strings into a constant and use that constant instead, I think there is no debate about it, is it? No matter if it is only one string within a class file, the recommendation would be move that to a constant (static analysis, etc). Same with this `java` plugin. To me, it is a plain constant that could be moved into version catalogs.
2. About the `java` plugin changing, think about the `kotlin` id, and the fact that it did indeed change, for example, `org.jetbrains.kotlin.jvm` is becoming the recommending id to use.

Now, my main struggle here is what do we do when I made that change, moving `java` into version catalogs, in some other project, would you come at me to move it back to where it was? 🤔 That's why I am trying here to align on it, and any other minor suggestion such as this, I don't mind what will happen here and now, but rather next and if the situation reverses.

PS: Thanks for the discussion, as always, but please ignore it, you don't have to act on it if you don't feel like you need to, I totally understand.

ParaskP7 commented on gradle/libs.versions.toml:
```diff
@@ -247,6 +245,7 @@ kotlin-jvm = { id = "org.jetbrains.kotlin.jvm", version.ref = "kotlin-main" }
 kotlin-parcelize = { id = "org.jetbrains.kotlin.plugin.parcelize", version.ref = "kotlin-main" }
 kotlin-serialization = { id = "org.jetbrains.kotlin.plugin.serialization", version.ref = "kotlin-main" }
 kotlinx-kover = { id = "org.jetbrains.kotlinx.kover", version.ref = "kotlinx-kover" }
+kapt = { id = "org.jetbrains.kotlin.kapt", version.ref = "kotlin-main" }
```
Minor (🔍): Alphabetically `kapt` should come before `kotlin-allopen`, consider moving it up a bit. 😊

wzieba replied:
👍 e78fceba4ff1d4b1219acad04efb642b1e737205

ParaskP7 commented on config/lint/lint.xml:
```diff
@@ -46,6 +46,7 @@
     <issue id="WeakPrng" severity="warning" />
     <issue id="TapjackingVulnerable" severity="warning" />
     <issue id="InsecurePermissionProtectionLevel" severity="warning" />
+    <issue id="CustomX509TrustManager" severity="warning" />
```
Suggestion (💡): Consider adding this `<!-- fluxc -->` comment next to this config so that we are aware that this was added due to the `fluxc` module (assuming this is because of that module and not the `login` or any other module).

PS: Same applies to all other config changes below: `VectorRaster` + `IconDensities` + `KaptUsageInsteadOfKsp`

wzieba replied:
> Suggestion (💡): Consider adding this <!-- fluxc --> comment

👍 [2e078a6](https://github.com/wordpress-mobile/WordPress-Android/pull/21261/commits/2e078a6dd25860b87818ca0dd8e9683e3a5aa658)

ParaskP7 commented on config/lint/lint.xml:
```diff
@@ -82,4 +83,8 @@
     <!--  We rely on more granural checks from WordPress-Lint-Android  -->
     <issue id="UnknownNullness" severity="ignore" />
+
+    <issue id="VectorRaster" severity="warning" />
```
Minor (🔍): Consider moving this under the `CORRECTNESS` group.

PS: Same applies to the other two:
1. `IconDensities` to `USABILITY` group.
2. `KaptUsageInsteadOfKsp` to `PERFORMANCE` group.

-----

Now that we are on it, I also noticed that:
1. `UnknownNullness` is defined twice, we should remove the first one.
2. `UsingMaterialAndMaterial3Libraries` should be also moved into the `CORRECTNESS` group.

This way this `INTEROPERABILITY` group will correct itself, having only the `UnknownNullness` config, as initially intended.

wzieba replied:
> Minor (🔍): Consider moving this under the CORRECTNESS group.

👍 [2e078a6](https://github.com/wordpress-mobile/WordPress-Android/pull/21261/commits/2e078a6dd25860b87818ca0dd8e9683e3a5aa658)

> Now that we are on it, I also noticed that:

True, but I'll keep this out of scope of this PR - I'll address it later!

ParaskP7 replied:
Awesome, thanks for this change! 👍

wzieba committed:
Revert not treating warnings as errors in the project

Disable warningsAsErrors in `fluxc` only. It was not enabled in `fluxc` before vendoring.

ParaskP7 commented on settings.gradle:
```diff
@@ -16,6 +16,36 @@ plugins {
+            version("apache-commons-text", "1.10.0")
+            version("facebook-flipper", "0.51.0")
+            version("facebook-soloader", "0.9.0")
+
```
Minor (🔍): Consider removing this empty line.

ParaskP7 commented on libs/login/gradle.properties-example:
```diff
@@ -0,0 +1,17 @@
+# Project-wide Gradle settings.
+
+# Optional: used to autofill username and password fields at login on debug build only
```
Suggestion (💡): Consider removing everything but these 4 `wp` related properties from this example `gradle.properties` file.

```properties
# Optional: used to autofill username and password fields at login on debug build only
wp.debug.wpcom_login_email =
wp.debug.wpcom_login_username =
wp.debug.wpcom_login_password =
wp.debug.wpcom_website_url =
```

FYI: The rest of the properties are anyway matching the parent `gradle.properties` file (`android.useAndroidX=true`, etc).

PS: Same applies to the example `gradle.properties` file on the `fluxc` level, the recommended change being:

```properties
# Optional: used to enable Wordpress API and database downgrade on debug build only
fluxc.ENABLE_WPAPI = false
wp.ENABLE_DATABASE_DOWNGRADE = false
```

Btw, I am not sure why this `gradle.properties-example` file is not shown on this diff, but it is anyway shown on my branch after me checking-out it via this PR, weird... 🤷

ParaskP7 replied:
FYI: This is done as part of #21353 and c85ec844790477331e08ee546a807acb6e6aaeb6.

ParaskP7 commented on .buildkite/pipeline.yml:
```diff
@@ -107,6 +107,26 @@ steps:
         artifact_paths:
           - "**/build/test-results/merged-test-results.xml"
 
+      - label: "🔬 Unit test FluxC"
```
Minor (🔍): For consistency purposes, this could be renamed to a capital `Test` instead, just to match the rest of them.

PS: Same applies to `Unit test Login` below.

wzieba replied:
👍 [c79408b](https://github.com/wordpress-mobile/WordPress-Android/pull/21261/commits/c79408bb3eba2e877a201c06d5083b079b687e8f)

ParaskP7 commented on .buildkite/pipeline.yml:
```diff
@@ -107,6 +107,26 @@ steps:
+        command: ".buildkite/commands/run-unit-tests.sh fluxc"
+        plugins:
+          - $CI_TOOLKIT
+          - $TEST_COLLECTOR:
```
Minor (🔍): I noticed an extra space on all the other `$TEST_COLLECTOR` steps, maybe this extra space could be removing from everywhere. 😊

ParaskP7 commented on libs/fluxc/build.gradle:
```diff
@@ -3,7 +3,7 @@ plugins {
     alias(sharedLibs.plugins.kotlin.android)
     alias(sharedLibs.plugins.kotlin.parcelize)
     alias(sharedLibs.plugins.kotlin.kapt)
```
Suggestion (💡): Since the `kotlin` version is now aligned (a13f97ee87ae671366300ed7ff5183cf68bbfd08), consider replacing `sharedLibs` with `libs` for everything `plugins.kotlin`, at least on the `plugins` side of things here.

PS.1: `android.library` is already matching so that can be replaced too, which leaves us with only `dependencies` on `sharedLibs` (for later).
PS.2: Kind of the same applies to the `login` module, better we replace `id` with `alias` from version catalogs.

ParaskP7 replied:
FYI: This is done as part of #21353 and 88a675f760b3baa8e10d173c339a040fdf209c87.

ParaskP7 reviewed (approved):
👋 @wzieba !

I have reviewed and (quickly) tested this PR as per the instructions, everything works as expected, an amazing job well done, kudos! 🌟 x 🌟 ^ 🌟 

-----

I have left few suggestions (💡) and some minor (🔍) comments for you to consider. I am going to approve this PR anyway, since none is blocking. I am **NOT** going to merge this PR yet to give you some time to apply any of my suggestions. However, feel free to ignore them and merge the PR yourself.

-----

`EXTRAS`
1. I think we can now remove the `Do Not Merge` label from this and the other two PRs (#21304 & #21266)
2. I didn't do a thorough smoke test of the project, just a quick one. I am hoping that, after you apply any of the comment included in this review, @nbradbury will jump on it with a thorough smoke test and give us a final 🟢 light for the merge.

wzieba committed:
Update lint config to match existing convention

wzieba committed:
Update names of unit test jobs to match convention

wzieba commented:
Thanks @ParaskP7 for the review! I'll minimize suggested changes to include only these which are about code added by this PR and don't touch vendored (or copy-pasted) code to keep things with minimal changes in this PR.

I'll address the rest of suggestions in a separate PR.

wzieba committed:
Reorder libs.versions.toml to match alphabetical order
